## Multi strategy time
### Scalping
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/scalping \
  --forward-start 2025-01-01 \
  --score-col "score_scalp__multi_strategy_time__rank_hgb__scalp" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.6 \
  --min-traded-value 500000000 \
  --require-broksum 

### Swing defensive `0.161`
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/swing_hgb_defensive \
  --forward-start 2025-01-01 \
  --score-col "score_swing__multi_strategy_time__hgb__swing" \
  --top-k 15 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly

### swing aggresive `0.485`
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/swing_hgb_aggresive \
  --forward-start 2025-01-01 \
  --score-col "score_swing__multi_strategy_time__xgb__swing" \
  --top-k 15 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 1000000000 \
  --require-broksum

### Position `0.117`
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/position \
  --forward-start 2025-01-01 \
  --score-col "score_position__multi_strategy_time__xgb__position" \
  --top-k 15 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 1000000000 \
  --require-broksum


## ARA predictor
### More safe and risk `0.511`
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/ara_candidate \
  --forward-start 2025-01-01 \
  --score-col "score_ara" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 1000000000 \
  --require-broksum

### Very high risk `0.870`
  python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/ara_watchlist_only \
  --forward-start 2025-01-01 \
  --score-col "score_ara__ara_predictor__xgb__ara_tomorrow" \
  --top-k 1 \
  --hold-days 5 \
  --min-score 0.55 \
  --min-traded-value 500000000 \
  --require-broksum


## Momentum ranker
### Momentum 5d defensif
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_5d_hgb_defensif \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_5d__momentum_ranker__xgb__momentum_5d" \
  --top-k 2 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 2000000000 \
  --require-broksum


### Momentum 5d aggresive
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_5d_hgb-aggresive \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_5d__momentum_ranker__regime_hgb__momentum_5d" \
  --top-k 7 \
  --hold-days 2 \
  --min-score 0.60 \
  --min-traded-value 500000000 \
  --require-broksum


### Momentum 10d defensif
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_ranker_10d_defensif \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_10d__momentum_ranker__xgb__momentum_10d" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 2000000000

### Momentum 10d aggresive
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_ranker_10d_aggresive \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_10d__momentum_ranker__hgb__momentum_10d" \
  --top-k 15 \
  --hold-days 2 \
  --min-score 0.60 \
  --min-traded-value 500000000 \
  --require-broksum


## Market Maker
### silent_accum_breakout defensif
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/market_maker_silent_accum_defensif \
  --forward-start 2025-01-01 \
  --score-col "score_mm_silent" \
  --top-k 7 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 1000000000 \
  --require-broksum


### silent_accum_breakout aggresive
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/market_maker_silent_accum_aggresive \
  --forward-start 2025-01-01 \
  --score-col "score_mm_silent__market_maker_accumulation__xgb__silent_accum_breakout" \
  --top-k 3 \
  --hold-days 2 \
  --min-score 0.50 \
  --min-traded-value 500000000 \
  --exclude-broker-value-anomaly


ARA == ARA CANDIDATE:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/ara_candidate \
  --forward-start 2025-01-01 \
  --score-col "score_ara" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 1000000000 \
  --require-broksum

MARKET MAKER == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/market_maker_silent_accum_defensif \
  --forward-start 2025-01-01 \
  --score-col "score_mm_silent" \
  --top-k 7 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 1000000000 \
  --require-broksum


MOMENTUM RANKER 5D == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_5d_hgb_defensif \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_5d__momentum_ranker__xgb__momentum_5d" \
  --top-k 2 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 2000000000 \
  --require-broksum

MOMENTUM RANKER 10D == AGGRESIVE:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/momentum_ranker_10d_aggresive \
  --forward-start 2025-01-01 \
  --score-col "score_momentum_10d__momentum_ranker__hgb__momentum_10d" \
  --top-k 15 \
  --hold-days 2 \
  --min-score 0.60 \
  --min-traded-value 500000000 \
  --require-broksum
  

MULTI STRATEGY SCALPING:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/scalping \
  --forward-start 2025-01-01 \
  --score-col "score_scalp__multi_strategy_time__rank_hgb__scalp" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.6 \
  --min-traded-value 500000000 \
  --require-broksum 

MULTI STRATEGY SWING == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/swing_hgb_defensive \
  --forward-start 2025-01-01 \
  --score-col "score_swing__multi_strategy_time__hgb__swing" \
  --top-k 15 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly

MULTI STRATEGY POSITION:
python3 workflows/run_forward_test.py \
  --panel signals/validation/validation_signal_panel.parquet \
  --output-dir data/evaluation/forward_tests/position \
  --forward-start 2025-01-01 \
  --score-col "score_position__multi_strategy_time__xgb__position" \
  --top-k 15 \
  --hold-days 1 \
  --min-score 0.55 \
  --min-traded-value 1000000000 \
  --require-broksum

# Monte carlo:
1. ARA
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/ara_candidate/trades.parquet \
  --output-dir data/evaluation/monte_carlo/ara_candidate \
  --n-iter 5000

2. Market Maker
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/market_maker_silent_accum_defensif/trades.parquet \
  --output-dir data/evaluation/monte_carlo/market_maker_silent_accum_defensif \
  --n-iter 5000

3. Momentum 5D
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/momentum_5d_hgb_defensif/trades.parquet \
  --output-dir data/evaluation/monte_carlo/momentum_5d_hgb_defensif \
  --n-iter 5000

4. Momentum 10D
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/momentum_ranker_10d_aggresive/trades.parquet \
  --output-dir data/evaluation/monte_carlo/momentum_ranker_10d_aggresive \
  --n-iter 5000

5. Scalping
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/scalping/trades.parquet \
  --output-dir data/evaluation/monte_carlo/scalping \
  --n-iter 5000

6. Swing
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/swing_hgb_defensive/trades.parquet \
  --output-dir data/evaluation/monte_carlo/swing_hgb_defensive \
  --n-iter 5000

7. Position
python3 workflows/run_monte_carlo.py \
  --trades data/evaluation/forward_tests/position/trades.parquet \
  --output-dir data/evaluation/monte_carlo/position \
  --n-iter 5000

# Rolling Back / Continual Learning `Backtest Search`

In [5]:
import polars as pl

df = pl.read_parquet("/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/features/live/latest/base_features.parquet")

recent = (
    df.filter(pl.col("date") >= pl.date(2026, 5, 19))
    .group_by("date")
    .agg([
        pl.len().alias("initial"),
        pl.col("close").is_not_null().sum().alias("valid_close"),
        ((pl.col("close") >= 100) & (pl.col("close") <= 1000)).sum().alias("price_100_1000"),
        (pl.col("traded_value_proxy").fill_null(0) >= 500_000_000).sum().alias("min_traded_value"),
        (pl.col("has_broksum").fill_null(0) == 1).sum().alias("has_broksum"),
        (
            (pl.col("close").is_not_null()) &
            (pl.col("close") >= 100) &
            (pl.col("close") <= 1000) &
            (pl.col("traded_value_proxy").fill_null(0) >= 500_000_000) &
            (pl.col("has_broksum").fill_null(0) == 1)
        ).sum().alias("final_eligible"),
    ])
    .sort("date")
)

print(recent)

shape: (2, 7)
┌────────────┬─────────┬─────────────┬───────────────┬───────────────┬─────────────┬───────────────┐
│ date       ┆ initial ┆ valid_close ┆ price_100_100 ┆ min_traded_va ┆ has_broksum ┆ final_eligibl │
│ ---        ┆ ---     ┆ ---         ┆ 0             ┆ lue           ┆ ---         ┆ e             │
│ date       ┆ u32     ┆ u32         ┆ ---           ┆ ---           ┆ u32         ┆ ---           │
│            ┆         ┆             ┆ u32           ┆ u32           ┆             ┆ u32           │
╞════════════╪═════════╪═════════════╪═══════════════╪═══════════════╪═════════════╪═══════════════╡
│ 2026-05-19 ┆ 959     ┆ 959         ┆ 516           ┆ 482           ┆ 863         ┆ 252           │
│ 2026-05-20 ┆ 959     ┆ 959         ┆ 522           ┆ 449           ┆ 858         ┆ 231           │
└────────────┴─────────┴─────────────┴───────────────┴───────────────┴─────────────┴───────────────┘


## ARA == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/ara_candidate \
  --forward-start 2026-01-20 \
  --score-col "score_ara" \
  --top-k 7 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly


## ARA == AGGRESIVE:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/ara_watchlist \
  --forward-start 2026-01-20 \
  --score-col "score_ara__ara_predictor__hgb__ara_tomorrow" \
  --top-k 1 \
  --hold-days 1 \
  --min-score 0.50 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly

## MOMENTUM RANKER 5D == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/momentum_5d_hgb_defensif \
  --forward-start 2026-01-20 \
  --score-col "score_momentum_5d__momentum_ranker__regime_hgb__momentum_5d" \
  --top-k 10 \
  --hold-days 2 \
  --min-traded-value 1000000000 \
  --require-broksum


## MOMENTUM RANKER 10D == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/momentum_ranker_10d_aggresive \
  --forward-start 2026-01-20 \
  --score-col "score_momentum_10d__momentum_ranker__hgb__momentum_10d" \
  --top-k 7 \
  --hold-days 2 \
  --min-traded-value 1000000000 

## MOMENTUM RANKER 20D == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/momentum_ranker_20d_aggresive \
  --forward-start 2026-01-20 \
  --score-col "score_momentum_20d__momentum_ranker__regime_hgb__momentum_20d" \
  --top-k 5 \
  --hold-days 1 \
  --min-traded-value 2000000000 \
  --require-broksum 





## MULTI STRATEGY SCALPING == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/scalping \
  --forward-start 2026-01-20 \
  --score-col "score_scalp" \
  --top-k 3 \
  --hold-days 2 \
  --min-score 0.65 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly


## MULTI STRATEGY SWING == DEFENSIF:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/swing_hgb_defensive \
  --forward-start 2026-01-20 \
  --score-col "score_swing__multi_strategy_time__hgb__swing" \
  --top-k 7 \
  --hold-days 1 \
  --min-traded-value 500000000 \
  --require-broksum \
  --exclude-broker-value-anomaly

## MULTI STRATEGY POSITION:
python3 workflows/run_forward_test.py \
  --panel signals/validation/continual_q3_2024_q1_2026_panel_enriched.parquet \
  --output-dir data/evaluation/forward_tests/continual_learning/position \
  --forward-start 2026-01-20 \
  --score-col "score_position__multi_strategy_time__xgb__position" \
  --top-k 7 \
  --hold-days 2 \
  --min-score 0.55 \
  --min-traded-value 2000000000 \
  --require-broksum 

In [4]:
import polars as pl

path = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/features/history/base_features.parquet"

df = pl.read_parquet(path)

print(df.shape)
print(df.head())
print(df.tail())
print(
    df.select([
        pl.col("date").min().alias("min_date"),
        pl.col("date").max().alias("max_date"),
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("duplicate_ticker_date"),
    ])
)

(1782675, 198)
shape: (5, 198)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ date       ┆ ticker ┆ rank1_buye ┆ rank1_sel ┆ … ┆ rank1_buy ┆ rank1_sel ┆ rank1_sel ┆ rank1_sel │
│ ---        ┆ ---    ┆ r          ┆ ler       ┆   ┆ _freq     ┆ l_val     ┆ l_lot     ┆ l_freq    │
│ date       ┆ str    ┆ ---        ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆ str        ┆ str       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64       │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 2016-01-04 ┆ AALI   ┆ DB         ┆ DB        ┆ … ┆ 53.0      ┆ 1.1233e10 ┆ 7086.0    ┆ 449.0     │
│ 2016-01-04 ┆ ABBA   ┆ LK         ┆ LK        ┆ … ┆ 1.0       ┆ 1.17e8    ┆ 26000.0   ┆ 1.0       │
│ 2016-01-04 ┆ ABDA   ┆ YU         ┆ YU        ┆ … ┆ 1.0       ┆ 750000.0  ┆ 1.0       ┆ 1.0       │
│ 2016-01-04 ┆ ABMM   ┆ YP         ┆ RF        ┆ … ┆ 2.0    

In [3]:
import pandas as pd

macro = pd.read_parquet("data/features/macro/macro_features.parquet")
macro.to_csv("mac.csv", index=False)

In [1]:
import polars as pl

m = pl.read_parquet("data/raw_canonical/macro.parquet")
mf = pl.read_parquet("data/features/macro/macro_features.parquet")

print(m.shape, m.columns[:30])
print(m.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.col("macro_missing_flag").fill_null(0).sum().alias("macro_missing_rows"),
]))

print(mf.shape, mf.columns[:50])
print(mf.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.col("macro_missing_flag").fill_null(0).sum().alias("macro_missing_rows"),
]))

(4158, 16) ['date', 'idr_usd', 'usd_idr', 'wti', 'brent', 'coal_proxy', 'ihsg', 'bi_rate', 'macro_missing_flag', 'coal_gap_flag', 'oil_avg', 'wti_return', 'brent_return', 'fx_return', 'coal_proxy_return', 'bi_rate_change']
shape: (1, 3)
┌────────────┬────────────┬────────────────────┐
│ min_date   ┆ max_date   ┆ macro_missing_rows │
│ ---        ┆ ---        ┆ ---                │
│ date       ┆ date       ┆ i64                │
╞════════════╪════════════╪════════════════════╡
│ 2015-01-01 ┆ 2026-05-20 ┆ 0                  │
└────────────┴────────────┴────────────────────┘
(4158, 74) ['date', 'idr_usd', 'usd_idr', 'wti', 'brent', 'coal_proxy', 'ihsg', 'bi_rate', 'macro_missing_flag', 'coal_gap_flag', 'oil_avg', 'wti_return', 'brent_return', 'fx_return', 'coal_proxy_return', 'bi_rate_change', 'market_ret_1d', 'idr_usd_ret_5d', 'idr_usd_vol_5d', 'brent_ret_5d', 'wti_ret_5d', 'oil_avg_ret_5d', 'brent_vol_5d', 'coal_ret_5d', 'coal_vol_5d', 'ihsg_ret_5d', 'idr_usd_ret_10d', 'idr_usd_vol_10d

# Check EOD Updates

## Apakah live feature store sudah punya 2026-05-21?

In [1]:
import polars as pl

df = pl.read_parquet("data/features/live/latest/base_features.parquet")

print(df.shape)
print(df.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup"),
]))

(60389, 198)
shape: (1, 3)
┌────────────┬────────────┬─────┐
│ min_date   ┆ max_date   ┆ dup │
│ ---        ┆ ---        ┆ --- │
│ date       ┆ date       ┆ u32 │
╞════════════╪════════════╪═════╡
│ 2026-02-23 ┆ 2026-06-05 ┆ 0   │
└────────────┴────────────┴─────┘


## Feature store sudah punya 2026-05-21?

In [3]:
import polars as pl

df = pl.read_parquet("data/features/live/latest/base_features.parquet")

recent = (
    df.filter(pl.col("date") >= pl.date(2026, 5, 19))
    .group_by("date")
    .agg([
        pl.len().alias("initial"),
        pl.col("close").is_not_null().sum().alias("valid_close"),
        ((pl.col("close") >= 100) & (pl.col("close") <= 1000)).sum().alias("price_100_1000"),
        (pl.col("traded_value_proxy").fill_null(0) >= 500_000_000).sum().alias("min_traded_value"),
        (pl.col("has_broksum").fill_null(0) == 1).sum().alias("has_broksum"),
        (
            (pl.col("close").is_not_null()) &
            (pl.col("close") >= 100) &
            (pl.col("close") <= 1000) &
            (pl.col("traded_value_proxy").fill_null(0) >= 500_000_000) &
            (pl.col("has_broksum").fill_null(0) == 1)
        ).sum().alias("final_eligible"),
    ])
    .sort("date")
)

print(recent)

shape: (8, 7)
┌────────────┬─────────┬─────────────┬───────────────┬───────────────┬─────────────┬───────────────┐
│ date       ┆ initial ┆ valid_close ┆ price_100_100 ┆ min_traded_va ┆ has_broksum ┆ final_eligibl │
│ ---        ┆ ---     ┆ ---         ┆ 0             ┆ lue           ┆ ---         ┆ e             │
│ date       ┆ u32     ┆ u32         ┆ ---           ┆ ---           ┆ u32         ┆ ---           │
│            ┆         ┆             ┆ u32           ┆ u32           ┆             ┆ u32           │
╞════════════╪═════════╪═════════════╪═══════════════╪═══════════════╪═════════════╪═══════════════╡
│ 2026-05-19 ┆ 959     ┆ 959         ┆ 516           ┆ 482           ┆ 863         ┆ 252           │
│ 2026-05-20 ┆ 959     ┆ 959         ┆ 522           ┆ 449           ┆ 858         ┆ 231           │
│ 2026-05-21 ┆ 959     ┆ 959         ┆ 506           ┆ 454           ┆ 863         ┆ 225           │
│ 2026-05-22 ┆ 959     ┆ 959         ┆ 508           ┆ 409           ┆ 861   

## Canonical raw sudah punya 21 May atau belum?

In [2]:
import polars as pl

for p in [
    "data/raw_canonical/ohlcv.parquet",
    "data/raw_canonical/broker_summary.parquet",
    "data/raw_canonical/insider_activity.parquet",
    "data/raw_canonical/neo_bdm.parquet",
    "data/raw_canonical/tradebook_price.parquet",
    "data/raw_canonical/macro.parquet",
    "data/raw_canonical/tradebook_time.parquet",
]:
    df = pl.read_parquet(p)
    print(p, df.shape, df.select([
        pl.col("date").min().alias("min"),
        pl.col("date").max().alias("max"),
    ]))

data/raw_canonical/ohlcv.parquet (1956373, 11) shape: (1, 2)
┌────────────┬────────────┐
│ min        ┆ max        │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2015-01-02 ┆ 2026-06-05 │
└────────────┴────────────┘
data/raw_canonical/broker_summary.parquet (23490538, 16) shape: (1, 2)
┌────────────┬────────────┐
│ min        ┆ max        │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2016-01-04 ┆ 2026-06-05 │
└────────────┴────────────┘
data/raw_canonical/insider_activity.parquet (92479, 8) shape: (1, 2)
┌────────────┬────────────┐
│ min        ┆ max        │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2017-10-12 ┆ 2026-06-05 │
└────────────┴────────────┘
data/raw_canonical/neo_bdm.parquet (7678, 13) shape: (1, 2)
┌────────────┬────────────┐
│ min        ┆ max        │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2026-01-2

In [5]:
import pandas as pd


df = pd.read_excel("data/pure_raw/trading_summary/RingkasanSaham-20260605.xlsx")
df.to_csv("lansa.csv", index=False)

In [ ]:
!ls -lt data/pure_raw/insider_activity/ | head -6

total 256
-rw-r--r--@ 1 albert  staff   3138 Jun  2 22:14 insider_activity_2026-05-28_to_2026-06-02_2026-06-02_22-14-38.csv
-rw-r--r--@ 1 albert  staff  35037 May 28 21:09 insider_activity_2026-05-21_to_2026-05-28_2026-05-28_21-09-47.csv
-rw-r--r--@ 1 albert  staff  29900 May 22 12:19 insider_activity_2026-05-14_to_2026-05-21_2026-05-22_12-19-18.csv
-rw-r--r--@ 1 albert  staff  52754 May 16 10:29 insider_activity_2026-05-04_to_2026-05-14_2026-05-16_10-29-30.csv
-rw-r--r--@ 1 albert  staff    609 May  5 14:38 insider_activity_2026-05-04_to_2026-05-04_2026-05-05_14-38-35.csv


In [4]:
import polars as pl

df = pl.read_parquet("data/features/live/latest/base_features.parquet")

print(df.shape)
print(df.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup"),
    pl.col("has_broksum").fill_null(0).sum().alias("broksum_rows"),
]))

(53676, 198)
shape: (1, 4)
┌────────────┬────────────┬─────┬──────────────┐
│ min_date   ┆ max_date   ┆ dup ┆ broksum_rows │
│ ---        ┆ ---        ┆ --- ┆ ---          │
│ date       ┆ date       ┆ u32 ┆ i64          │
╞════════════╪════════════╪═════╪══════════════╡
│ 2026-02-23 ┆ 2026-05-29 ┆ 0   ┆ 46549        │
└────────────┴────────────┴─────┴──────────────┘


# cek feature parity history vs live

In [1]:
import polars as pl

hist = pl.read_parquet("data/features/history/base_features.parquet")
live = pl.read_parquet("data/features/live/latest/base_features.parquet")

hist_cols = set(hist.columns)
live_cols = set(live.columns)

print("history:", hist.shape)
print("live   :", live.shape)

print("only in history:", sorted(hist_cols - live_cols), len(hist_cols - live_cols))
print("only in live   :", sorted(live_cols - hist_cols), len(live_cols - hist_cols))

print(hist.select([
    pl.col("date").min().alias("hist_min"),
    pl.col("date").max().alias("hist_max"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("hist_dup"),
]))

print(live.select([
    pl.col("date").min().alias("live_min"),
    pl.col("date").max().alias("live_max"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("live_dup"),
]))

history: (1782675, 198)
live   : (39301, 198)
only in history: [] 0
only in live   : [] 0
shape: (1, 3)
┌────────────┬────────────┬──────────┐
│ hist_min   ┆ hist_max   ┆ hist_dup │
│ ---        ┆ ---        ┆ ---      │
│ date       ┆ date       ┆ u32      │
╞════════════╪════════════╪══════════╡
│ 2016-01-04 ┆ 2025-12-30 ┆ 0        │
└────────────┴────────────┴──────────┘
shape: (1, 3)
┌────────────┬────────────┬──────────┐
│ live_min   ┆ live_max   ┆ live_dup │
│ ---        ┆ ---        ┆ ---      │
│ date       ┆ date       ┆ u32      │
╞════════════╪════════════╪══════════╡
│ 2026-03-09 ┆ 2026-05-20 ┆ 0        │
└────────────┴────────────┴──────────┘


In [27]:
import polars as pl

hist = pl.read_parquet("data/features/history/base_features.parquet")
live = pl.read_parquet("data/features/live/latest/base_features.parquet")

hist_cols = set(hist.columns)
live_cols = set(live.columns)

print("history:", hist.shape)
print("live   :", live.shape)

print("only in history:", sorted(hist_cols - live_cols), len(hist_cols - live_cols))
print("only in live   :", sorted(live_cols - hist_cols), len(live_cols - hist_cols))

print(hist.select([
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("hist_dup"),
    pl.col("macro_missing_flag").fill_null(0).sum().alias("hist_macro_missing"),
    pl.col("usd_idr").is_not_null().sum().alias("hist_usd_idr_non_null"),
    pl.col("brent").is_not_null().sum().alias("hist_brent_non_null"),
    pl.col("bi_rate").is_not_null().sum().alias("hist_bi_rate_non_null"),
]))

print(live.select([
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("live_dup"),
    pl.col("macro_missing_flag").fill_null(0).sum().alias("live_macro_missing"),
    pl.col("usd_idr").is_not_null().sum().alias("live_usd_idr_non_null"),
    pl.col("brent").is_not_null().sum().alias("live_brent_non_null"),
    pl.col("bi_rate").is_not_null().sum().alias("live_bi_rate_non_null"),
]))

history: (1782675, 198)
live   : (36424, 198)
only in history: [] 0
only in live   : [] 0
shape: (1, 5)
┌──────────┬────────────────────┬──────────────────────┬─────────────────────┬─────────────────────┐
│ hist_dup ┆ hist_macro_missing ┆ hist_usd_idr_non_nul ┆ hist_brent_non_null ┆ hist_bi_rate_non_nu │
│ ---      ┆ ---                ┆ l                    ┆ ---                 ┆ ll                  │
│ u32      ┆ i64                ┆ ---                  ┆ u32                 ┆ ---                 │
│          ┆                    ┆ u32                  ┆                     ┆ u32                 │
╞══════════╪════════════════════╪══════════════════════╪═════════════════════╪═════════════════════╡
│ 0        ┆ 0                  ┆ 1782675              ┆ 1782675             ┆ 1782675             │
└──────────┴────────────────────┴──────────────────────┴─────────────────────┴─────────────────────┘
shape: (1, 5)
┌──────────┬────────────────────┬──────────────────────┬──────────────────

## Broker Ratio Sanity Check

In [28]:
import polars as pl

df = pl.read_parquet("data/features/history/base_features.parquet")

for c in ["net_flow_ratio", "buyer_dominance_ratio", "seller_dominance_ratio"]:
    print(c)
    print(df.select([
        pl.col(c).is_nan().sum().alias("nan"),
        pl.col(c).is_infinite().sum().alias("inf"),
        pl.col(c).is_null().sum().alias("null"),
        pl.col(c).min().alias("min"),
        pl.col(c).max().alias("max"),
    ]))

net_flow_ratio
shape: (1, 5)
┌─────┬─────┬────────┬───────────┬─────┐
│ nan ┆ inf ┆ null   ┆ min       ┆ max │
│ --- ┆ --- ┆ ---    ┆ ---       ┆ --- │
│ u32 ┆ u32 ┆ u32    ┆ f64       ┆ f64 │
╞═════╪═════╪════════╪═══════════╪═════╡
│ 0   ┆ 0   ┆ 396527 ┆ -0.125568 ┆ 1.0 │
└─────┴─────┴────────┴───────────┴─────┘
buyer_dominance_ratio
shape: (1, 5)
┌─────┬─────┬────────┬──────────┬─────┐
│ nan ┆ inf ┆ null   ┆ min      ┆ max │
│ --- ┆ --- ┆ ---    ┆ ---      ┆ --- │
│ u32 ┆ u32 ┆ u32    ┆ f64      ┆ f64 │
╞═════╪═════╪════════╪══════════╪═════╡
│ 0   ┆ 0   ┆ 396527 ┆ 0.043718 ┆ 1.0 │
└─────┴─────┴────────┴──────────┴─────┘
seller_dominance_ratio
shape: (1, 5)
┌─────┬─────┬────────┬──────────┬─────┐
│ nan ┆ inf ┆ null   ┆ min      ┆ max │
│ --- ┆ --- ┆ ---    ┆ ---      ┆ --- │
│ u32 ┆ u32 ┆ u32    ┆ f64      ┆ f64 │
╞═════╪═════╪════════╪══════════╪═════╡
│ 0   ┆ 0   ┆ 420279 ┆ 0.045119 ┆ 1.0 │
└─────┴─────┴────────┴──────────┴─────┘


# Cek training target distribution

In [31]:
import polars as pl

df = pl.read_parquet("data/datasets/training/full_labeled.parquet")

targets = [
    "label_scalp",
    "label_swing",
    "label_position",
    "label_brok_cont",
    "label_ara_tomorrow",
    "label_silent_accum_breakout",
    "label_big_runner_30d",
]

print(df.shape)

for c in targets:
    print("\n", c, c in df.columns)
    if c in df.columns:
        print(df.group_by(c).len().sort(c))

(1782675, 214)

 label_scalp True
shape: (3, 2)
┌─────────────┬─────────┐
│ label_scalp ┆ len     │
│ ---         ┆ ---     │
│ i8          ┆ u32     │
╞═════════════╪═════════╡
│ null        ┆ 38219   │
│ 0           ┆ 1430121 │
│ 1           ┆ 314335  │
└─────────────┴─────────┘

 label_swing True
shape: (3, 2)
┌─────────────┬─────────┐
│ label_swing ┆ len     │
│ ---         ┆ ---     │
│ i8          ┆ u32     │
╞═════════════╪═════════╡
│ null        ┆ 45187   │
│ 0           ┆ 1422488 │
│ 1           ┆ 315000  │
└─────────────┴─────────┘

 label_position True
shape: (3, 2)
┌────────────────┬─────────┐
│ label_position ┆ len     │
│ ---            ┆ ---     │
│ i8             ┆ u32     │
╞════════════════╪═════════╡
│ null           ┆ 54929   │
│ 0              ┆ 1421077 │
│ 1              ┆ 306669  │
└────────────────┴─────────┘

 label_brok_cont True
shape: (3, 2)
┌─────────────────┬─────────┐
│ label_brok_cont ┆ len     │
│ ---             ┆ ---     │
│ i8              ┆ u32    

In [32]:
import polars as pl

df = pl.read_parquet("data/datasets/training/full_labeled.parquet")

targets = [
    "label_scalp",
    "label_swing",
    "label_position",
    "label_brok_cont",
    "label_ara_tomorrow",
    "label_silent_accum_breakout",
    "label_big_runner_30d",
]

print("shape:", df.shape)

for c in targets:
    print("\n" + c)
    print("exists:", c in df.columns)
    if c in df.columns:
        print(df.group_by(c).len().sort(c))
        print(
            df.select(
                (pl.col(c).fill_null(0).sum() / pl.len()).alias("positive_rate")
            )
        )

shape: (1782675, 214)

label_scalp
exists: True
shape: (3, 2)
┌─────────────┬─────────┐
│ label_scalp ┆ len     │
│ ---         ┆ ---     │
│ i8          ┆ u32     │
╞═════════════╪═════════╡
│ null        ┆ 38219   │
│ 0           ┆ 1430121 │
│ 1           ┆ 314335  │
└─────────────┴─────────┘
shape: (1, 1)
┌───────────────┐
│ positive_rate │
│ ---           │
│ f64           │
╞═══════════════╡
│ 0.176328      │
└───────────────┘

label_swing
exists: True
shape: (3, 2)
┌─────────────┬─────────┐
│ label_swing ┆ len     │
│ ---         ┆ ---     │
│ i8          ┆ u32     │
╞═════════════╪═════════╡
│ null        ┆ 45187   │
│ 0           ┆ 1422488 │
│ 1           ┆ 315000  │
└─────────────┴─────────┘
shape: (1, 1)
┌───────────────┐
│ positive_rate │
│ ---           │
│ f64           │
╞═══════════════╡
│ 0.176701      │
└───────────────┘

label_position
exists: True
shape: (3, 2)
┌────────────────┬─────────┐
│ label_position ┆ len     │
│ ---            ┆ ---     │
│ i8             ┆ u

## Cek target momentum

In [33]:
import polars as pl

df = pl.read_parquet("data/datasets/training/full_labeled.parquet")

targets = [
    "label_momentum_5d",
    "label_momentum_10d",
    "label_momentum_20d",
]

for c in targets:
    print("\n", c, c in df.columns)
    if c in df.columns:
        print(df.group_by(c).len().sort(c))


 label_momentum_5d True
shape: (3, 2)
┌───────────────────┬─────────┐
│ label_momentum_5d ┆ len     │
│ ---               ┆ ---     │
│ i8                ┆ u32     │
╞═══════════════════╪═════════╡
│ null              ┆ 40232   │
│ 0                 ┆ 1423664 │
│ 1                 ┆ 318779  │
└───────────────────┴─────────┘

 label_momentum_10d True
shape: (3, 2)
┌────────────────────┬─────────┐
│ label_momentum_10d ┆ len     │
│ ---                ┆ ---     │
│ i8                 ┆ u32     │
╞════════════════════╪═════════╡
│ null               ┆ 45187   │
│ 0                  ┆ 1418438 │
│ 1                  ┆ 319050  │
└────────────────────┴─────────┘

 label_momentum_20d True
shape: (3, 2)
┌────────────────────┬─────────┐
│ label_momentum_20d ┆ len     │
│ ---                ┆ ---     │
│ i8                 ┆ u32     │
╞════════════════════╪═════════╡
│ null               ┆ 54929   │
│ 0                  ┆ 1410618 │
│ 1                  ┆ 317128  │
└────────────────────┴─────────┘

python3 workflows/train_models.py \
  --root . \
  --fold-set yearly \
  --families multi_strategy_time,sm_tracker,ara_predictor,market_maker_accumulation \
  --algos lgb,xgb,hgb,rank_hgb,regime_hgb

## ?

In [26]:
import polars as pl

hist = pl.read_parquet("data/features/history/base_features.parquet")
live = pl.read_parquet("data/features/live/latest/base_features.parquet")

for name, df in [("history", hist), ("live", live)]:
    print(name, df.shape)
    print(df.select([
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup_ticker_date"),
        pl.col("macro_missing_flag").fill_null(0).sum().alias("macro_missing_rows"),
        pl.col("usd_idr").is_not_null().sum().alias("usd_idr_non_null"),
        pl.col("brent").is_not_null().sum().alias("brent_non_null"),
        pl.col("bi_rate").is_not_null().sum().alias("bi_rate_non_null"),
    ]))

history (1782675, 115)
shape: (1, 5)
┌─────────────────┬────────────────────┬──────────────────┬────────────────┬──────────────────┐
│ dup_ticker_date ┆ macro_missing_rows ┆ usd_idr_non_null ┆ brent_non_null ┆ bi_rate_non_null │
│ ---             ┆ ---                ┆ ---              ┆ ---            ┆ ---              │
│ u32             ┆ i64                ┆ u32              ┆ u32            ┆ u32              │
╞═════════════════╪════════════════════╪══════════════════╪════════════════╪══════════════════╡
│ 0               ┆ 0                  ┆ 1782675          ┆ 1782675        ┆ 1782675          │
└─────────────────┴────────────────────┴──────────────────┴────────────────┴──────────────────┘
live (36424, 145)
shape: (1, 5)
┌─────────────────┬────────────────────┬──────────────────┬────────────────┬──────────────────┐
│ dup_ticker_date ┆ macro_missing_rows ┆ usd_idr_non_null ┆ brent_non_null ┆ bi_rate_non_null │
│ ---             ┆ ---                ┆ ---              ┆ ---    

## Check fold n=20

In [2]:
import polars as pl

path = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/datasets/folds/quarterly/fold_19_train.parquet"

df = pl.read_parquet(path)

print(df.shape)
print(df.head())
print(
    df.select([
        pl.col("date").min().alias("min_date"),
        pl.col("date").max().alias("max_date"),
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("duplicate_ticker_date"),
    ])
)

(1642880, 224)
shape: (5, 224)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ date       ┆ ticker ┆ rank1_buye ┆ rank1_sel ┆ … ┆ label_mul ┆ label_mom ┆ label_mom ┆ label_mom │
│ ---        ┆ ---    ┆ r          ┆ ler       ┆   ┆ tibagger_ ┆ entum_5d  ┆ entum_10d ┆ entum_20d │
│ date       ┆ str    ┆ ---        ┆ ---       ┆   ┆ 60d       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆ str        ┆ str       ┆   ┆ ---       ┆ i8        ┆ i8        ┆ i8        │
│            ┆        ┆            ┆           ┆   ┆ i8        ┆           ┆           ┆           │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 2016-01-04 ┆ AALI   ┆ DB         ┆ DB        ┆ … ┆ 0         ┆ 0         ┆ 1         ┆ 0         │
│ 2016-01-04 ┆ ABBA   ┆ LK         ┆ LK        ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 2016-01-04 ┆ ABDA   ┆ YU         ┆ YU        ┆ … ┆ 0      

In [4]:
import polars as pl

path = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/datasets/folds/quarterly/fold_20_train.parquet"

df = pl.read_parquet(path)

print(df.shape)
print(df.head())
print(
    df.select([
        pl.col("date").min().alias("min_date"),
        pl.col("date").max().alias("max_date"),
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("duplicate_ticker_date"),
    ])
)

(1700434, 224)
shape: (5, 224)
┌────────────┬────────┬────────────┬───────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ date       ┆ ticker ┆ rank1_buye ┆ rank1_sel ┆ … ┆ label_mul ┆ label_mom ┆ label_mom ┆ label_mom │
│ ---        ┆ ---    ┆ r          ┆ ler       ┆   ┆ tibagger_ ┆ entum_5d  ┆ entum_10d ┆ entum_20d │
│ date       ┆ str    ┆ ---        ┆ ---       ┆   ┆ 60d       ┆ ---       ┆ ---       ┆ ---       │
│            ┆        ┆ str        ┆ str       ┆   ┆ ---       ┆ i8        ┆ i8        ┆ i8        │
│            ┆        ┆            ┆           ┆   ┆ i8        ┆           ┆           ┆           │
╞════════════╪════════╪════════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ 2016-01-04 ┆ AALI   ┆ DB         ┆ DB        ┆ … ┆ 0         ┆ 0         ┆ 1         ┆ 0         │
│ 2016-01-04 ┆ ABBA   ┆ LK         ┆ LK        ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 0         │
│ 2016-01-04 ┆ ABDA   ┆ YU         ┆ YU        ┆ … ┆ 0      

# Cek validational signal panel

In [6]:
import polars as pl

p = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/signals/validation/validation_signal_panel.parquet"
df = pl.read_parquet(p)

df

date,ticker,score_sm__sm_tracker__hgb__brok_cont,score_sm__sm_tracker__rank_hgb__brok_cont,score_sm__sm_tracker__regime_hgb__brok_cont,score_sm__sm_tracker__xgb__brok_cont,score_ara__ara_predictor__hgb__ara_tomorrow,score_ara_near__ara_predictor__hgb__near_ara_tomorrow,score_ara__ara_predictor__regime_hgb__ara_tomorrow,score_ara_near__ara_predictor__regime_hgb__near_ara_tomorrow,score_ara__ara_predictor__xgb__ara_tomorrow,score_ara_near__ara_predictor__xgb__near_ara_tomorrow,score_mm_big__market_maker_accumulation__hgb__big_runner_30d,score_mm_silent__market_maker_accumulation__hgb__silent_accum_breakout,score_mm_big__market_maker_accumulation__regime_hgb__big_runner_30d,score_mm_silent__market_maker_accumulation__regime_hgb__silent_accum_breakout,score_mm_big__market_maker_accumulation__xgb__big_runner_30d,score_mm_silent__market_maker_accumulation__xgb__silent_accum_breakout,score_sm,score_ara,score_ara_near,score_mm_big,score_mm_silent,close,traded_value_proxy,has_broksum,broker_value_anomaly_flag,rank1_buyer,fwd_ret_1d,fwd_ret_2d,fwd_ret_3d,fwd_ret_5d,fwd_ret_10d,fwd_ret_20d,fwd_ret_30d
date,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,i8,i8,str,f64,f64,f64,f64,f64,f64,f64
2021-01-04,"""AALI""",0.166219,0.359086,0.337114,0.334225,0.018081,0.012969,0.014249,0.014448,0.01815,0.007348,0.101729,0.000816,0.136016,0.001038,0.223474,0.001502,0.166219,0.014249,0.012969,0.223474,0.000816,13075.0,9.9535e10,1,0,"""YP""",-0.005736,-0.024857,-0.007648,-0.028681,-0.110899,-0.120459,-0.156788
2021-01-04,"""ABBA""",0.311443,0.553758,0.496837,0.484938,0.041505,0.011888,0.040972,0.011173,0.032986,0.014801,0.311626,0.360206,0.231032,0.489219,0.311048,0.369046,0.311443,0.040972,0.011888,0.311048,0.360206,80.0,2.8786e9,1,0,"""YP""",-0.025,-0.0375,-0.025,-0.0375,-0.0125,-0.2625,-0.2
2021-01-04,"""ABDA""",0.012219,0.038113,0.043047,0.029125,0.012329,0.005103,0.017036,0.005597,0.010296,0.004915,0.248324,0.000018,0.176919,0.000026,0.138746,0.000067,0.012219,0.017036,0.005103,0.138746,0.000018,5575.0,0.0,0,null,null,0.0,0.0,0.0,0.004484,0.040359,0.058296,0.09417
2021-01-04,"""ABMM""",0.1068,0.208499,0.221794,0.199686,0.038455,0.015231,0.038551,0.013362,0.039452,0.017519,0.121397,0.21618,0.165034,0.213326,0.220189,0.214094,0.1068,0.038551,0.015231,0.220189,0.21618,780.0,1.55298e8,1,0,"""PD""",-0.019231,-0.038462,-0.019231,0.070513,-0.032051,-0.089744,-0.032051
2021-01-04,"""ACES""",0.040155,0.103714,0.109384,0.103471,0.002411,0.006283,0.003517,0.004959,0.006284,0.004814,0.276412,0.000799,0.188397,0.000735,0.308892,0.001319,0.040155,0.003517,0.006283,0.308892,0.000799,1700.0,2.0726e10,1,0,"""KZ""",0.035294,0.017647,0.047059,0.035294,0.032353,-0.055882,-0.044118
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
2025-12-30,"""WOOD""",null,null,null,null,null,null,null,null,null,null,null,0.001534,null,0.001981,null,0.001883,null,null,null,null,0.001534,378.0,1.2952e9,1,0,"""XL""",null,null,null,null,null,null,null
2025-12-30,"""WSBP""",null,null,null,null,null,null,null,null,null,null,null,0.001776,null,0.002024,null,0.004459,null,null,null,null,0.001776,23.0,8.712722e8,1,0,"""XL""",null,null,null,null,null,null,null
2025-12-30,"""WTON""",null,null,null,null,null,null,null,null,null,null,null,0.001627,null,0.002034,null,0.001199,null,null,null,null,0.001627,101.0,5.9489e9,1,0,"""CP""",null,null,null,null,null,null,null


In [1]:
import polars as pl

p = "signals/validation/sm_tracker_panel.parquet"
df = pl.read_parquet(p)

print(df.shape)
print([c for c in df.columns if c.startswith("score_")])
print(
    df.select([
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup_ticker_date")
    ])
)

(1024340, 19)
['score_sm__sm_tracker__hgb__brok_cont', 'score_sm__sm_tracker__rank_hgb__brok_cont', 'score_sm__sm_tracker__regime_hgb__brok_cont', 'score_sm__sm_tracker__xgb__brok_cont', 'score_sm']
shape: (1, 1)
┌─────────────────┐
│ dup_ticker_date │
│ ---             │
│ u32             │
╞═════════════════╡
│ 0               │
└─────────────────┘


# Check Build live base scores

In [8]:
import json
import polars as pl

df = pl.read_parquet("signals/live/base_scores_2026-05-13.parquet")
policy = json.load(open("configs/signal_policy.json"))

needed = []
for group in ["core_execution", "aggressive_challenger", "watchlist_only"]:
    for s in policy.get(group, []):
        needed.append(s["score_col"])

print("base scores:", df.shape)
print("missing:", [c for c in needed if c not in df.columns])
print([c for c in df.columns if c.startswith("score_")])

base scores: (229, 206)
missing: []
['score_sm', 'score_ara', 'score_mm_silent', 'score_momentum_5d__momentum_ranker__xgb__momentum_5d', 'score_momentum_10d__momentum_ranker__hgb__momentum_10d', 'score_scalp__multi_strategy_time__rank_hgb__scalp', 'score_swing__multi_strategy_time__hgb__swing', 'score_position__multi_strategy_time__xgb__position']


In [5]:
import polars as pl

df = pl.read_parquet("signals/live/base_scores_2026-05-19.parquet")

print(df.shape)
print([c for c in df.columns if c.startswith("score_")])

(252, 206)
['score_sm', 'score_ara', 'score_mm_silent', 'score_momentum_5d__momentum_ranker__xgb__momentum_5d', 'score_momentum_10d__momentum_ranker__hgb__momentum_10d', 'score_scalp__multi_strategy_time__rank_hgb__scalp', 'score_swing__multi_strategy_time__hgb__swing', 'score_position__multi_strategy_time__xgb__position']


# Check hasil build continual feature store

In [7]:
import polars as pl

df = pl.read_parquet("data/features/continual/q3_2024_to_2026_05_20/base_features.parquet")

print(df.shape)
print(df.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup"),
]))

(423912, 198)
shape: (1, 3)
┌────────────┬────────────┬─────┐
│ min_date   ┆ max_date   ┆ dup │
│ ---        ┆ ---        ┆ --- │
│ date       ┆ date       ┆ u32 │
╞════════════╪════════════╪═════╡
│ 2024-07-01 ┆ 2026-05-20 ┆ 0   │
└────────────┴────────────┴─────┘


In [8]:
import polars as pl

df = pl.read_parquet("data/datasets/continual/q3_2024_to_2026_04_30/full_labeled.parquet")

targets = [
    "label_scalp",
    "label_swing",
    "label_position",
    "label_brok_cont",
    "label_ara_tomorrow",
    "label_silent_accum_breakout",
    "label_big_runner_30d",
    "label_momentum_5d",
    "label_momentum_10d",
    "label_momentum_20d",
]

print("shape:", df.shape)

for c in targets:
    print("\n", c, c in df.columns)
    if c in df.columns:
        print(df.group_by(c).len().sort(c))

shape: (416225, 208)

 label_scalp True
shape: (2, 2)
┌─────────────┬────────┐
│ label_scalp ┆ len    │
│ ---         ┆ ---    │
│ i8          ┆ u32    │
╞═════════════╪════════╡
│ 0           ┆ 335279 │
│ 1           ┆ 80946  │
└─────────────┴────────┘

 label_swing True
shape: (3, 2)
┌─────────────┬────────┐
│ label_swing ┆ len    │
│ ---         ┆ ---    │
│ i8          ┆ u32    │
╞═════════════╪════════╡
│ null        ┆ 3896   │
│ 0           ┆ 332119 │
│ 1           ┆ 80210  │
└─────────────┴────────┘

 label_position True
shape: (3, 2)
┌────────────────┬────────┐
│ label_position ┆ len    │
│ ---            ┆ ---    │
│ i8             ┆ u32    │
╞════════════════╪════════╡
│ null           ┆ 18502  │
│ 0              ┆ 319531 │
│ 1              ┆ 78192  │
└────────────────┴────────┘

 label_brok_cont True
shape: (3, 2)
┌─────────────────┬────────┐
│ label_brok_cont ┆ len    │
│ ---             ┆ ---    │
│ i8              ┆ u32    │
╞═════════════════╪════════╡
│ null            

## cek panel signal

In [1]:
import polars as pl

p = "signals/validation/continual_q3_2024_q1_2026_panel.parquet"
df = pl.read_parquet(p)

print(df.shape)
print([c for c in df.columns if c.startswith("score_")])
print(df.select([
    pl.col("date").min().alias("min_date"),
    pl.col("date").max().alias("max_date"),
    pl.struct(["date", "ticker"]).is_duplicated().sum().alias("dup_ticker_date"),
]))

df

(298682, 49)
['score_ara__ara_predictor__hgb__ara_tomorrow', 'score_ara__ara_predictor__rank_hgb__ara_tomorrow', 'score_ara__ara_predictor__regime_hgb__ara_tomorrow', 'score_ara__ara_predictor__xgb__ara_tomorrow', 'score_momentum_10d__momentum_ranker__hgb__momentum_10d', 'score_momentum_20d__momentum_ranker__hgb__momentum_20d', 'score_momentum_5d__momentum_ranker__hgb__momentum_5d', 'score_momentum_10d__momentum_ranker__rank_hgb__momentum_10d', 'score_momentum_20d__momentum_ranker__rank_hgb__momentum_20d', 'score_momentum_5d__momentum_ranker__rank_hgb__momentum_5d', 'score_momentum_10d__momentum_ranker__regime_hgb__momentum_10d', 'score_momentum_20d__momentum_ranker__regime_hgb__momentum_20d', 'score_momentum_5d__momentum_ranker__regime_hgb__momentum_5d', 'score_momentum_10d__momentum_ranker__xgb__momentum_10d', 'score_momentum_20d__momentum_ranker__xgb__momentum_20d', 'score_momentum_5d__momentum_ranker__xgb__momentum_5d', 'score_position__multi_strategy_time__hgb__position', 'score_s

date,ticker,score_ara__ara_predictor__hgb__ara_tomorrow,score_ara__ara_predictor__rank_hgb__ara_tomorrow,score_ara__ara_predictor__regime_hgb__ara_tomorrow,score_ara__ara_predictor__xgb__ara_tomorrow,score_momentum_10d__momentum_ranker__hgb__momentum_10d,score_momentum_20d__momentum_ranker__hgb__momentum_20d,score_momentum_5d__momentum_ranker__hgb__momentum_5d,score_momentum_10d__momentum_ranker__rank_hgb__momentum_10d,score_momentum_20d__momentum_ranker__rank_hgb__momentum_20d,score_momentum_5d__momentum_ranker__rank_hgb__momentum_5d,score_momentum_10d__momentum_ranker__regime_hgb__momentum_10d,score_momentum_20d__momentum_ranker__regime_hgb__momentum_20d,score_momentum_5d__momentum_ranker__regime_hgb__momentum_5d,score_momentum_10d__momentum_ranker__xgb__momentum_10d,score_momentum_20d__momentum_ranker__xgb__momentum_20d,score_momentum_5d__momentum_ranker__xgb__momentum_5d,score_position__multi_strategy_time__hgb__position,score_scalp__multi_strategy_time__hgb__scalp,score_swing__multi_strategy_time__hgb__swing,score_position__multi_strategy_time__rank_hgb__position,score_scalp__multi_strategy_time__rank_hgb__scalp,score_swing__multi_strategy_time__rank_hgb__swing,score_position__multi_strategy_time__regime_hgb__position,score_scalp__multi_strategy_time__regime_hgb__scalp,score_swing__multi_strategy_time__regime_hgb__swing,score_position__multi_strategy_time__xgb__position,score_scalp__multi_strategy_time__xgb__scalp,score_swing__multi_strategy_time__xgb__swing,score_ara,score_momentum_10d,score_momentum_20d,score_momentum_5d,score_position,score_scalp,score_swing,close,traded_value_proxy,has_broksum,broker_value_anomaly_flag,rank1_buyer,fwd_ret_1d,fwd_ret_2d,fwd_ret_3d,fwd_ret_5d,fwd_ret_10d,fwd_ret_20d,fwd_ret_30d
date,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,i8,i8,str,f64,f64,f64,f64,f64,f64,f64
2025-01-02,"""AADI""",0.002102,0.00382,0.006135,0.002107,0.227062,0.049228,0.162454,0.132578,0.075522,0.159343,0.136297,0.077558,0.209011,0.152819,0.101611,0.177207,0.049228,0.295164,0.162454,0.075522,0.272008,0.159343,0.077558,0.272195,0.209011,0.101611,0.253096,0.177207,0.002107,0.152819,0.101611,0.177207,0.101611,0.253096,0.177207,8225.0,1.7398e11,1,0,"""XL""",-0.045593,-0.066869,-0.042553,-0.06383,0.069909,0.075988,-0.018237
2025-01-02,"""AALI""",0.011286,0.007279,0.005547,0.005582,0.079669,0.054305,0.110389,0.119778,0.066439,0.138985,0.124831,0.056319,0.124538,0.102673,0.119402,0.135006,0.054305,0.190122,0.110389,0.066439,0.155784,0.138985,0.056319,0.165798,0.124538,0.119402,0.170314,0.135006,0.005582,0.102673,0.119402,0.135006,0.119402,0.170314,0.135006,6150.0,1.0412e9,1,0,"""CC""",0.004065,-0.00813,-0.012195,-0.028455,-0.028455,-0.060976,-0.036585
2025-01-02,"""ABBA""",0.031312,0.043778,0.043411,0.020402,0.345937,0.364836,0.270816,0.362879,0.327975,0.310363,0.349192,0.382835,0.317549,0.333957,0.352246,0.316238,0.364836,0.304518,0.270816,0.327975,0.280965,0.310363,0.382835,0.305154,0.317549,0.352246,0.247568,0.316238,0.020402,0.333957,0.352246,0.316238,0.352246,0.247568,0.316238,27.0,2.56338e7,1,0,"""CC""",0.0,0.0,0.0,0.0,0.037037,-0.296296,-0.259259
2025-01-02,"""ABDA""",0.011408,0.009523,0.007971,0.007694,0.051778,0.090163,0.067639,0.063259,0.120149,0.071286,0.065209,0.099309,0.090384,0.069827,0.114865,0.118214,0.090163,0.065032,0.067639,0.120149,0.055766,0.071286,0.099309,0.037026,0.090384,0.114865,0.051865,0.118214,0.007694,0.069827,0.114865,0.118214,0.114865,0.051865,0.118214,3800.0,0.0,0,null,null,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2025-01-02,"""ABMM""",0.001596,0.004338,0.003922,0.003165,0.06845,0.085547,0.160225,0.105921,0.123511,0.145367,0.111362,0.088647,0.155801,0.089345,0.111372,0.129149,0.085547,0.124934,0.160225,0.123511,0.163112,0.145367,0.088647,0.135258,0.155801,0.111372,0.153482,0.129149,0.003165,0.089345,0.111372,0.129149,0.111372,0.153482,0.129149,3570.0,4.3558e9,1,0,"""CC""",0.002801,0.002801,-0.011204

# BI RATE Scraper

In [22]:
# CONFIG
api_key = "cdc65504d0b176f0b763f2d8dcd855dd"

import requests
import time
import json
import os
import random

api_key = ""

years = {
    2026: 126,
    2025: 125,
    2024: 124,
    2023: 123,
    2022: 122,
    2021: 121,
    2020: 120,
    2019: 119,
    2018: 118,
    2017: 117,
    2016: 116,
    2015: 115,
}

base_url = "https://webapi.bps.go.id/v1/api/list/model/data/lang/eng/domain/0000/var/379/th/{th}/key/{key}"

output_dir = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate"
os.makedirs(output_dir, exist_ok=True)

log_file = os.path.join(output_dir, "download_log.txt")

# FUNCTION: FETCH WITH RETRY
def fetch_bps(th, api_key, max_retry=3):
    url = base_url.format(th=th, key=api_key)
    for attempt in range(1, max_retry + 1):
        try:
            r = requests.get(url, timeout=30)
            # sukses
            if r.status_code == 200:
                return r.json()
            print(f"HTTP {r.status_code} (attempt {attempt})")
        except Exception as e:
            print(f"Exception (attempt {attempt}): {e}")
        # exponential backoff
        sleep_time = (2 ** attempt) + random.random()
        print(f"RETRY in {sleep_time:.2f}s")
        time.sleep(sleep_time)
    raise Exception(f"Failed after {max_retry} retries for th={th}")

# MAIN LOOP
for year, th in years.items():
    file_path = os.path.join(output_dir, f"bi_rate_{year}.json")
    # RESUME LOGIC (skip jika sudah ada)
    if os.path.exists(file_path):
        print(f"SKIP {year} (already exists)")
        continue
    print(f"\n FETCHING {year}...")
    try:
        data = fetch_bps(th, api_key)
        # validasi status API
        if data.get("status") != "OK":
            print(f"!!! {year} status NOT OK")
            continue
        # simpan file
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"SAVED {file_path}")
        # log success
        with open(log_file, "a") as log:
            log.write(f"SUCCESS {year}\n")
    except Exception as e:
        print(f"FAILED {year}: {e}")
        # log failure
        with open(log_file, "a") as log:
            log.write(f"FAILED {year}: {e}\n")

    # SAFE DELAY (anti overload)
    sleep_time = 1.5 + random.random()
    print(f"SLEEP {sleep_time:.2f}s\n")
    time.sleep(sleep_time)

print("\nDONE - All available data processed")

<>:83: SyntaxWarning: invalid escape sequence '\D'
<>:83: SyntaxWarning: invalid escape sequence '\D'
/var/folders/pt/n4kz04ns6wn8sv6msgx6mqb80000gn/T/ipykernel_7933/3722219837.py:83: SyntaxWarning: invalid escape sequence '\D'
  print("\DONE - All available data processed")



 FETCHING 2026...
HTTP 500 (attempt 1)
⏳ retry in 2.94s
SAVED /Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate/bi_rate_2026.json
💤 sleep 2.48s


 FETCHING 2025...
SAVED /Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate/bi_rate_2025.json
💤 sleep 1.93s


 FETCHING 2024...
SAVED /Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate/bi_rate_2024.json
💤 sleep 2.23s


 FETCHING 2023...
HTTP 500 (attempt 1)
⏳ retry in 2.90s
HTTP 500 (attempt 2)
⏳ retry in 4.88s
SAVED /Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate/bi_rate_2023.json
💤 sleep 1.78s


 FETCHING 2022...
HTTP 500 (attempt 1)
⏳ retry in 2.28s
HTTP 500 (attempt 2)
⏳ retry in 4.53s
SAVED /Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/bps_bi_rate/bi_rate_2022.json
💤 sleep 2.49s


 FETCHING 2021...
SAVED /Users/albert/D

In [ ]:
# python3 workflows/build_macro.py \
#   --root . \
#   --start-date 2015-01-01 \
#   --end-date 2026-05-13 \
#   --mode scrape \
#   --force \
#   --download-bi-rate \
#   --bps-api-key "cdc65504d0b176f0b763f2d8dcd855dd"

# python3 workflows/build_canonical_raw.py \
#   --root . \
#   --start-date 2015-01-01 \
#   --end-date 2026-05-13 \
#   --macro-mode scrape \
#   --force-macro \
#   --download-bi-rate \
#   --bps-api-key "cdc65504d0b176f0b763f2d8dcd855dd"

# Feature coverage QA

In [10]:
import polars as pl

df = pl.read_parquet("data/features/history/base_features.parquet")

cols = [
    "has_broksum",
    "buy_val_total_sane",
    "net_flow_ratio",
    "rank1_same_buyer_streak",
    "rank1_buyer_daily_share",
    "has_insider_activity",
    "has_corporate_action",
    "macro_missing_flag",
    "has_bdm_any",
]

for c in cols:
    if c in df.columns:
        print("\n", c)
        print(
            df.select([
                pl.len().alias("rows"),
                pl.col(c).is_not_null().sum().alias("non_null"),
                pl.col(c).fill_null(0).sum().alias("sum_or_count"),
            ])
        )
    else:
        print("MISSING:", c)


 has_broksum
shape: (1, 3)
┌─────────┬──────────┬──────────────┐
│ rows    ┆ non_null ┆ sum_or_count │
│ ---     ┆ ---      ┆ ---          │
│ u32     ┆ u32      ┆ i64          │
╞═════════╪══════════╪══════════════╡
│ 1782675 ┆ 1782675  ┆ 1386175      │
└─────────┴──────────┴──────────────┘

 buy_val_total_sane
shape: (1, 3)
┌─────────┬──────────┬──────────────┐
│ rows    ┆ non_null ┆ sum_or_count │
│ ---     ┆ ---      ┆ ---          │
│ u32     ┆ u32      ┆ f64          │
╞═════════╪══════════╪══════════════╡
│ 1782675 ┆ 1329367  ┆ 2.1043e16    │
└─────────┴──────────┴──────────────┘

 net_flow_ratio
shape: (1, 3)
┌─────────┬──────────┬──────────────┐
│ rows    ┆ non_null ┆ sum_or_count │
│ ---     ┆ ---      ┆ ---          │
│ u32     ┆ u32      ┆ f64          │
╞═════════╪══════════╪══════════════╡
│ 1782675 ┆ 1360424  ┆ NaN          │
└─────────┴──────────┴──────────────┘

 rank1_same_buyer_streak
shape: (1, 3)
┌─────────┬──────────┬──────────────┐
│ rows    ┆ non_null ┆ sum_or_

In [12]:
import polars as pl

paths = [
    "data/features/history/base_features.parquet",
    "data/features/live/latest/base_features.parquet",
]

for p in paths:
    df = pl.read_parquet(p)
    numeric_cols = [
        c for c, d in zip(df.columns, df.dtypes)
        if d in (pl.Float32, pl.Float64)
    ]

    df = df.with_columns([
        pl.when(pl.col(c).is_nan() | pl.col(c).is_infinite())
          .then(None)
          .otherwise(pl.col(c))
          .alias(c)
        for c in numeric_cols
    ])

    df.write_parquet(p)
    print("cleaned", p, df.shape)

cleaned data/features/history/base_features.parquet (1782675, 138)
cleaned data/features/live/latest/base_features.parquet (39301, 138)


In [13]:
import polars as pl

df = pl.read_parquet("data/features/history/base_features.parquet")

for c in ["net_flow_ratio", "buyer_dominance_ratio", "seller_dominance_ratio"]:
    if c in df.columns:
        print(c)
        print(df.select([
            pl.col(c).is_nan().sum().alias("nan"),
            pl.col(c).is_infinite().sum().alias("inf"),
            pl.col(c).is_null().sum().alias("null"),
            pl.col(c).min().alias("min"),
            pl.col(c).max().alias("max"),
        ]))

net_flow_ratio
shape: (1, 5)
┌─────┬─────┬────────┬────────────────┬──────────┐
│ nan ┆ inf ┆ null   ┆ min            ┆ max      │
│ --- ┆ --- ┆ ---    ┆ ---            ┆ ---      │
│ u32 ┆ u32 ┆ u32    ┆ f64            ┆ f64      │
╞═════╪═════╪════════╪════════════════╪══════════╡
│ 0   ┆ 0   ┆ 426330 ┆ -145085.894059 ┆ 6.0189e6 │
└─────┴─────┴────────┴────────────────┴──────────┘
buyer_dominance_ratio
shape: (1, 5)
┌─────┬─────┬────────┬──────────┬─────┐
│ nan ┆ inf ┆ null   ┆ min      ┆ max │
│ --- ┆ --- ┆ ---    ┆ ---      ┆ --- │
│ u32 ┆ u32 ┆ u32    ┆ f64      ┆ f64 │
╞═════╪═════╪════════╪══════════╪═════╡
│ 0   ┆ 0   ┆ 396500 ┆ 0.043718 ┆ 1.0 │
└─────┴─────┴────────┴──────────┴─────┘
seller_dominance_ratio
shape: (1, 5)
┌─────┬─────┬────────┬──────────┬─────┐
│ nan ┆ inf ┆ null   ┆ min      ┆ max │
│ --- ┆ --- ┆ ---    ┆ ---      ┆ --- │
│ u32 ┆ u32 ┆ u32    ┆ f64      ┆ f64 │
╞═════╪═════╪════════╪══════════╪═════╡
│ 0   ┆ 0   ┆ 420253 ┆ 0.045119 ┆ 1.0 │
└─────┴─────┴────────

untuk OHLCV, actual coverage dari date=2015-01-02.

date,ticker,open,high,low,close,volume
2015-01-02,AALI,16715.333882176663,17196.84584211245,16680.940170752678,16904.5,1385945.0
2015-01-02,ABMM,2110.177001953125,2110.177001953125,2110.177001953125,2110.177001953125,0.0
ETC.


Broker Summary, actual coverage dari date=2016-01-04

stock_code,date,from,to,rank,BY,BY_type,B.val,B.lot,B.freq,B.avg,SL,SL_type,S.val,S.lot,S.freq,S.avg
AALI,20160104,2016-01-04,2016-01-04,1,DB,,6.55022e+09,4105,53,15956.686967113277,DB,,1.123298e+10,7086,449,15852.356759808072
AALI,20160104,2016-01-04,2016-01-04,2,FG,,4.4667825e+09,2846,287,15694.949051300071,NI,Pemerintah,1.259055e+09,801,87,15718.539325842698
ETC.


insider activity, actual coverage=12 Oct 17
8823,12 Oct 17,ASJT,DANA PENSIUN PERKEBUNAN,SELL,"1,200",0.00,"464,453,200",77.41,"464,452,000",77.41,0,,UNSPECIFIED,LOCAL,KSEI,Sumber: KSEI
8809,12 Oct 17,ACES,SMALLCAP WORLD FUND,BUY,"4,646,900",+0.03,"874,736,465",5.10,"879,383,365",5.13,0,,UNSPECIFIED,FOREIGN,KSEI,Sumber: KSEI

corporate action, actual coverage date=22 May 1979
20 Dec 1979,BATI,IPO,6600000,6600000
22 May 1979,CNTX,IPO,116000,116000

macro, expected coverage since 2015-01-02 and load IDR/UIDRTOUSD, COAL, BRENT, Etc.

neo bdm, actual coverage date=28_jan_2026
├── 28_jan_2026_neobdm-f-d.csv
├── 28_jan_2026_neobdm-f-w.csv
├── 28_jan_2026_neobdm-m-d.csv
├── 28_jan_2026_neobdm-m-w.csv
├── 28_jan_2026_neobdm-nr-d.csv
├── 28_jan_2026_neobdm-nr-w.csv


tradebook/orderbook, actual coverage date=2026-05-07
├── orderbook_levels_2026-05-04_21-42-12.csv
├── orderbook_snapshot_2026-05-04_21-42-12.csv
├── tradebook_price_2026-04-19_11-04-14.csv
├── tradebook_time_2026-04-19_11-04-14.csv


ketika saya menjalankan takap ke-2 yakni build feature store, ada masalah yang yang terjadi. Dimana historicall build berhasil[2.1] dieksekusi namun live feature tidak [2.2], anda bisa melakukan diagnosa serta melakukan analisa dan audit mendalam setelah melihat log dan reporr berikut:
2.1 build historical feature store
python3 workflows/build_feature_store.py \
  --root . \
  --scope history \
  --start-date 2016-01-01 \
  --end-date 2025-12-31
{
  "scope": "history", "start_date": "2016-01-01", "end_date": "2025-12-31", "rows": 1782675, "cols": 34,
  "path": "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/features/history/base_features.parquet",
  "feature_count": 32,
  "feature_cols": ["open", "high", "low", "close", "volume", "value", "frequency", "foreign_buy", "foreign_sell", "ret_1d", "ret_5d", "ret_10d", "ret_20d", "ma_5", "ma_20", "volume_ma20", "volatility_20d", "close_vs_ma20", "volume_ratio_20d", "traded_value_proxy", "insider_event_count", "insider_buy_count", "insider_sell_count", "insider_net_shares", "insider_net_pct_sum", "has_insider_activity", "has_broksum", "has_bdm_any", "cs_rank_ret_5d", "cs_rank_ret_10d", "cs_rank_ret_20d", "cs_rank_volume_ratio_20d"
  ]
}

saya juga mengecek base_features yang dihasilkan, berikut adalah hasilny:
(1782675, 34)
shape: (5, 34)

 date        ticker  open        high       …  cs_rank_r  cs_rank_r  cs_rank_r  cs_rank_v 
 ---         ---     ---         ---           et_5d      et_10d     et_20d     olume_rat 
 date        str     f64         f64           ---        ---        ---        io_20d    
                                               f64        f64        f64        ---       
                                                                                f64       

 2016-01-04  AALI    11127.7366  11197.943  …  null       null       null       null      
                     29          211                                                      
 2016-01-04  ABBA    39.726902   39.726902  …  null       null       null       null      
 2016-01-04  ABDA    6746.96679  6746.9667  …  null       null       null       null      
                     7           97                                                       
 2016-01-04  ABMM    1868.02568  1868.0256  …  null       null       null       null      
                     5           85                                                       
 2016-01-04  ACES    605.644027  616.92932  …  null       null       null       null      

shape: (1, 3)

 min_date    max_date    duplicate_ticker_date 
 ---         ---         ---                   
 date        date        u32                   

 2016-01-04  2025-12-30  0                     



2.2 build live feature store
python3 workflows/build_feature_store.py \
  --root . \
  --scope live \
  --start-date 2026-03-09 \
  --end-date 2026-05-13
Traceback (most recent call last):
  File "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/workflows/build_feature_store.py", line 12, in <module>
    if __name__=="__main__": main()
                             ^^^^^^
  File "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/workflows/build_feature_store.py", line 11, in main
    m=build_feature_store(root,a.start_date,a.end_date,a.scope,out); print(json.dumps(m,indent=2,default=str))
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/src/alpha_research/features/store.py", line 209, in build_feature_store
    base=add_price_features(load_ohlcv(raw_dir,start,end))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/src/alpha_research/features/store.py", line 39, in load_ohlcv
    raise ValueError(
ValueError: No OHLCV rows remain after date filtering. start=2026-09-03, end=2026-05-13, files_read=10, path=/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/raw/ohlcv. For historical feature store, make sure historical OHLCV files exist in data/raw/ohlcv, not only daily pure_raw trading summary files.

In [ ]:
import polars as pl

path = "/Users/albert/Documents/Finances/projects/02_alpha_research/alpha_research/data/news/event_intelligence/report/event_study_by_event_type.csv"

df = pl.read_parquet(path)

print(df.shape)
print(df.head())
print(
    df.select([
        pl.col("date").min().alias("min_date"),
        pl.col("date").max().alias("max_date"),
        pl.struct(["date", "ticker"]).is_duplicated().sum().alias("duplicate_ticker_date"),
    ])
)

(1995448, 11)
shape: (5, 11)
┌────────────┬────────┬─────────────┬────────────┬───┬───────┬───────────┬────────────┬────────────┐
│ date       ┆ ticker ┆ open        ┆ high       ┆ … ┆ value ┆ frequency ┆ foreign_bu ┆ foreign_se │
│ ---        ┆ ---    ┆ ---         ┆ ---        ┆   ┆ ---   ┆ ---       ┆ y          ┆ ll         │
│ date       ┆ str    ┆ f64         ┆ f64        ┆   ┆ f64   ┆ f64       ┆ ---        ┆ ---        │
│            ┆        ┆             ┆            ┆   ┆       ┆           ┆ f64        ┆ f64        │
╞════════════╪════════╪═════════════╪════════════╪═══╪═══════╪═══════════╪════════════╪════════════╡
│ 2015-01-02 ┆ AALI   ┆ 16715.33388 ┆ 17196.8458 ┆ … ┆ null  ┆ null      ┆ null       ┆ null       │
│            ┆        ┆ 2           ┆ 42         ┆   ┆       ┆           ┆            ┆            │
│ 2015-01-02 ┆ ABBA   ┆ 48.46682    ┆ 50.055897  ┆ … ┆ null  ┆ null      ┆ null       ┆ null       │
│ 2015-01-02 ┆ ABDA   ┆ 5533.175293 ┆ 5533.17529 ┆ … ┆ null  ┆

# Audit sample event type

In [1]:
import polars as pl

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

(
    df.filter(pl.col("event_type") == "suspension_uma")
      .select(["news_date", "source", "title", "ticker", "event_type", "event_side"])
      .head(30)
      .write_csv("data/news/event_intelligence/audit_suspension_uma_sample.csv")
)

## Buat QA report

In [2]:
import polars as pl

path = "data/news/event_intelligence/news_event_intelligence_dataset.parquet"
df = pl.read_parquet(path)

print("shape:", df.shape)

print("\nDate range:")
print(df.select([
    pl.col("news_date").min().alias("min_date"),
    pl.col("news_date").max().alias("max_date"),
]))

print("\nRows by year:")
print(
    df.with_columns(pl.col("news_date").dt.year().alias("year"))
      .group_by("year")
      .len()
      .sort("year")
)

print("\nSource distribution:")
print(df.group_by("source").len().sort("len", descending=True))

print("\nEvent type distribution:")
print(df.group_by("event_type").len().sort("len", descending=True))

print("\nEvent type by source:")
print(
    df.group_by(["source", "event_type"])
      .len()
      .sort("len", descending=True)
      .head(50)
)

print("\nOutcome completeness:")
outcome_cols = [
    "fwd_ret_5d",
    "market_alpha_5d",
    "sector_alpha_5d",
    "volatility_shock_5d",
    "mae_5d",
    "mfe_5d",
    "reaction_label_5d",
]
print(
    df.select([
        pl.col(c).is_null().mean().alias(f"{c}_null_rate")
        for c in outcome_cols
        if c in df.columns
    ])
)

print("\nDuplicate event_row_id:")
print(df.select(pl.col("event_row_id").n_unique().alias("unique_event_row_id")))
print("rows:", df.height)

shape: (127123, 122)

Date range:
shape: (1, 2)
┌────────────┬────────────┐
│ min_date   ┆ max_date   │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2020-01-02 ┆ 2026-05-23 │
└────────────┴────────────┘

Rows by year:
shape: (7, 2)
┌──────┬───────┐
│ year ┆ len   │
│ ---  ┆ ---   │
│ i32  ┆ u32   │
╞══════╪═══════╡
│ 2020 ┆ 2760  │
│ 2021 ┆ 14425 │
│ 2022 ┆ 20850 │
│ 2023 ┆ 18090 │
│ 2024 ┆ 25352 │
│ 2025 ┆ 26783 │
│ 2026 ┆ 18863 │
└──────┴───────┘

Source distribution:
shape: (3, 2)
┌─────────────┬───────┐
│ source      ┆ len   │
│ ---         ┆ ---   │
│ str         ┆ u32   │
╞═════════════╪═══════╡
│ IDX Channel ┆ 93472 │
│ unknown     ┆ 32939 │
│ Kabar Bursa ┆ 712   │
└─────────────┴───────┘

Event type distribution:
shape: (17, 2)
┌──────────────────────┬───────┐
│ event_type           ┆ len   │
│ ---                  ┆ ---   │
│ str                  ┆ u32   │
╞══════════════════════╪═══════╡
│ suspension_uma       ┆ 52386 │
│ market_commen

In [4]:
import polars as pl

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

print("shape:", df.shape)

print("\nDate range:")
print(df.select([
    pl.col("news_date").min().alias("min_date"),
    pl.col("news_date").max().alias("max_date"),
]))

print("\nEvent type distribution:")
print(
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
)

print("\nMacro currency sample titles:")
print(
    df.filter(pl.col("event_type") == "macro_currency")
      .select(["news_date", "title", "ticker", "event_type", "event_side"])
      .head(50)
)

shape: (127123, 122)

Date range:
shape: (1, 2)
┌────────────┬────────────┐
│ min_date   ┆ max_date   │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2020-01-02 ┆ 2026-05-23 │
└────────────┴────────────┘

Event type distribution:
shape: (21, 2)
┌───────────────────────────┬───────┐
│ event_type                ┆ len   │
│ ---                       ┆ ---   │
│ str                       ┆ u32   │
╞═══════════════════════════╪═══════╡
│ market_commentary         ┆ 32108 │
│ earnings_fundamental      ┆ 26333 │
│ business_development      ┆ 10352 │
│ dividend                  ┆ 8810  │
│ dilution_corporate_action ┆ 7551  │
│ …                         ┆ …     │
│ macro_rate                ┆ 1383  │
│ debt_financing            ┆ 1291  │
│ macro_currency            ┆ 1033  │
│ credit_rating             ┆ 726   │
│ stock_split               ┆ 551   │
└───────────────────────────┴───────┘

Macro currency sample titles:
shape: (50, 5)
┌────────────┬─────────

### Re-chek distribusi event type

In [6]:
import polars as pl

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

print(
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
)

shape: (17, 2)
┌──────────────────────┬───────┐
│ event_type           ┆ len   │
│ ---                  ┆ ---   │
│ str                  ┆ u32   │
╞══════════════════════╪═══════╡
│ earnings_fundamental ┆ 24319 │
│ market_commentary    ┆ 20979 │
│ accumulation_flow    ┆ 16589 │
│ dividend             ┆ 10846 │
│ distribution_flow    ┆ 9485  │
│ …                    ┆ …     │
│ general_news         ┆ 1735  │
│ macro_currency       ┆ 1269  │
│ macro_rate           ┆ 1093  │
│ geopolitical_risk    ┆ 430   │
│ index_rebalancing    ┆ 226   │
└──────────────────────┴───────┘


In [1]:
import polars as pl

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

print(
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
)

shape: (21, 2)
┌───────────────────────────┬───────┐
│ event_type                ┆ len   │
│ ---                       ┆ ---   │
│ str                       ┆ u32   │
╞═══════════════════════════╪═══════╡
│ market_commentary         ┆ 32189 │
│ earnings_fundamental      ┆ 24985 │
│ business_development      ┆ 11060 │
│ dividend                  ┆ 8810  │
│ dilution_corporate_action ┆ 7551  │
│ …                         ┆ …     │
│ macro_currency            ┆ 1524  │
│ macro_rate                ┆ 1361  │
│ debt_financing            ┆ 1291  │
│ credit_rating             ┆ 726   │
│ stock_split               ┆ 551   │
└───────────────────────────┴───────┘


In [3]:
import polars as pl

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

print(
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
)

shape: (21, 2)
┌───────────────────────────┬───────┐
│ event_type                ┆ len   │
│ ---                       ┆ ---   │
│ str                       ┆ u32   │
╞═══════════════════════════╪═══════╡
│ market_commentary         ┆ 32108 │
│ earnings_fundamental      ┆ 26333 │
│ business_development      ┆ 10352 │
│ dividend                  ┆ 8810  │
│ dilution_corporate_action ┆ 7551  │
│ …                         ┆ …     │
│ macro_rate                ┆ 1383  │
│ debt_financing            ┆ 1291  │
│ macro_currency            ┆ 1033  │
│ credit_rating             ┆ 726   │
│ stock_split               ┆ 551   │
└───────────────────────────┴───────┘


## Audit taxonomy event

In [3]:
import polars as pl
from pathlib import Path

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")
out = Path("data/news/event_intelligence/audit_samples")
out.mkdir(parents=True, exist_ok=True)

event_types = (
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
      .get_column("event_type")
      .to_list()
)

for et in event_types:
    safe = et.replace("/", "_").replace(" ", "_")
    (
        df.filter(pl.col("event_type") == et)
          .select([
              "news_date",
              "source",
              "news_category",
              "title",
              "ticker",
              "company_name",
              "sector",
              "event_type",
              "event_side",
              "impact_channel",
              "materiality_score",
              "uncertainty_score",
          ])
          .head(50)
          .write_csv(out / f"sample_{safe}.csv")
    )

print(f"Saved audit samples to: {out}")

Saved audit samples to: data/news/event_intelligence/audit_samples


### sample audit after rebuild

In [7]:
import polars as pl
from pathlib import Path

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

out = Path("data/news/event_intelligence/audit_samples_after_fix")
out.mkdir(parents=True, exist_ok=True)

for et in df.group_by("event_type").len().sort("len", descending=True)["event_type"].to_list():
    safe = et.replace("/", "_").replace(" ", "_")
    (
        df.filter(pl.col("event_type") == et)
          .select([
              "news_date",
              "source",
              "news_category",
              "title",
              "ticker",
              "company_name",
              "sector",
              "event_type",
              "event_side",
              "impact_channel",
              "materiality_score",
              "uncertainty_score",
          ])
          .head(50)
          .write_csv(out / f"sample_{safe}.csv")
    )

print("saved:", out)

saved: data/news/event_intelligence/audit_samples_after_fix


In [2]:
import polars as pl
from pathlib import Path

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

out = Path("data/news/event_intelligence/audit_samples_after_v37")
out.mkdir(parents=True, exist_ok=True)

event_types = (
    df.group_by("event_type")
      .len()
      .sort("len", descending=True)
      .get_column("event_type")
      .to_list()
)

for et in event_types:
    safe = et.replace("/", "_").replace(" ", "_")
    (
        df.filter(pl.col("event_type") == et)
          .select([
              "news_date",
              "source",
              "news_category",
              "title",
              "ticker",
              "company_name",
              "sector",
              "event_type",
              "event_side",
              "impact_channel",
              "materiality_score",
              "uncertainty_score",
          ])
          .head(50)
          .write_csv(out / f"sample_{safe}.csv")
    )

print("saved:", out)

saved: data/news/event_intelligence/audit_samples_after_v37


In [ ]:
import polars as pl
from pathlib import Path

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")

out = Path("data/news/event_intelligence/audit_samples_after_v38")
out.mkdir(parents=True, exist_ok=True)

for et in df.group_by("event_type").len().sort("len", descending=True)["event_type"].to_list():
    safe = et.replace("/", "_").replace(" ", "_")
    (
        df.filter(pl.col("event_type") == et)
          .select([
              "news_date",
              "source",
              "news_category",
              "title",
              "ticker",
              "company_name",
              "sector",
              "event_type",
              "event_side",
              "impact_channel",
              "materiality_score",
              "uncertainty_score",
          ])
          .head(50)
          .write_csv(out / f"sample_{safe}.csv")
    )

print("saved:", out)

## Live feature view dan label view

In [5]:
import polars as pl
from pathlib import Path

base = Path("data/news/event_intelligence")
df = pl.read_parquet(base / "news_event_intelligence_dataset.parquet")

id_cols = [
    "event_row_id", "article_id", "news_date", "entry_date", "ticker",
]

metadata_cols = [
    "source", "news_category",
    "sector", "subsector", "industry", "subindustry", "listing_board",
]

event_cols = [
    "event_scope", "event_type", "event_side", "impact_channel",
    "materiality_label", "materiality_score",
    "news_intensity_score", "uncertainty_score", "novelty_score",
    "bullish_keyword_hits", "bearish_keyword_hits", "uncertainty_keyword_hits",
    "same_event_seen_so_far",
]

live_market_cols = [
    "open", "high", "low", "close", "volume", "value", "frequency",
    "foreign_buy", "foreign_sell", "daily_ret", "is_zero_volume",
    "avg_volume_20d", "volume_ratio", "bwd_volatility_20d", "drawdown_20d",
]

bwd_cols = [
    c for c in df.columns
    if c.startswith("bwd_ret_") or c.startswith("bwd_volume_ratio_")
]

live_cols = [
    c for c in id_cols + metadata_cols + event_cols + live_market_cols + bwd_cols
    if c in df.columns
]

label_cols = [
    c for c in df.columns
    if (
        c.startswith("fwd_ret_")
        or c.startswith("ihsg_fwd_ret_")
        or c.startswith("sector_fwd_ret_")
        or c.startswith("market_alpha_")
        or c.startswith("sector_alpha_")
        or c.startswith("volatility_shock_")
        or c.startswith("mae_")
        or c.startswith("mfe_")
        or c.endswith("_label_5d")
        or c == "reaction_label_5d"
    )
]

label_view_cols = [c for c in id_cols + label_cols if c in df.columns]

live = df.select(live_cols)
labels = df.select(label_view_cols)

live.write_parquet(base / "news_event_live_features.parquet")
labels.write_parquet(base / "news_event_training_labels.parquet")

live.head(10).write_csv(base / "sample_live_features_10.csv")
labels.head(10).write_csv(base / "sample_training_labels_10.csv")

print("Full dataset:", df.shape)
print("Live features:", live.shape)
print("Training labels:", labels.shape)

Full dataset: (127123, 122)
Live features: (127123, 52)
Training labels: (127123, 63)


# Event study detail

In [6]:
import polars as pl
from pathlib import Path

df = pl.read_parquet("data/news/event_intelligence/news_event_intelligence_dataset.parquet")
out = Path("data/news/event_intelligence/report")
out.mkdir(parents=True, exist_ok=True)

def event_study(group_cols, filename):
    report = (
        df.group_by(group_cols)
          .agg([
              pl.len().alias("n_rows"),
              pl.col("article_id").n_unique().alias("n_articles"),
              pl.col("ticker").n_unique().alias("n_tickers"),

              pl.col("sector_alpha_5d").mean().alias("avg_sector_alpha_5d"),
              pl.col("sector_alpha_5d").median().alias("median_sector_alpha_5d"),
              (pl.col("sector_alpha_5d") > 0).mean().alias("hit_rate_sector_alpha_5d"),
              (pl.col("sector_alpha_5d") > 0.01).mean().alias("hit_rate_sector_alpha_gt_1pct_5d"),

              pl.col("market_alpha_5d").mean().alias("avg_market_alpha_5d"),
              (pl.col("market_alpha_5d") > 0).mean().alias("hit_rate_market_alpha_5d"),

              pl.col("volatility_shock_5d").mean().alias("avg_volatility_shock_5d"),
              (pl.col("volatility_shock_5d") > 1.5).mean().alias("volatility_shock_rate_5d"),

              pl.col("mae_5d").mean().alias("avg_mae_5d"),
              pl.col("mfe_5d").mean().alias("avg_mfe_5d"),

              (pl.col("sentiment_trap_label_5d") == 1).mean().alias("sentiment_trap_rate_5d"),
              (pl.col("delayed_reaction_label_5d") == 1).mean().alias("delayed_reaction_rate_5d"),
              (pl.col("acceleration_trigger_label_5d") == 1).mean().alias("acceleration_trigger_rate_5d"),
          ])
          .filter(pl.col("n_rows") >= 30)
          .sort("n_rows", descending=True)
    )
    report.write_csv(out / filename)
    return report

event_study(["event_type"], "event_study_by_event_type_detailed.csv")
event_study(["event_type", "event_side"], "event_study_by_event_type_side.csv")
event_study(["event_type", "sector"], "event_study_by_event_type_sector.csv")
event_study(["event_type", "news_category"], "event_study_by_event_type_category.csv")

print("saved to", out)

saved to data/news/event_intelligence/report
